lib/queryContextCache.js

```diff
@@ -0,0 +1,391 @@
+  this.context = context;
+};
+
+QueryContextElement.prototype.equals = function (obj) {
+  if (!(obj instanceof QueryContextElement)) {
+    return false;
+  }
+    
+  return (this.id === obj.id
+        && this.timestamp === obj.timestamp
+        && this.priority === obj.priority
+        && this.context.equals(obj.context));
+};
+
+QueryContextElement.prototype.hashCode = function () {
+  let hash = 31;
+  hash = hash * 31 + parseInt(this.id);
+  hash += (hash * 31) + parseInt(this.timestamp);
+  hash += (hash * 31) + parseInt(this.priority);
+  hash += (hash * 31) + parseInt(crypto.createHash('md5').update(Buffer.from(this.context,'utf-8').toString()).digest(),2);
+
+  return hash;
+};
+
+/**
+ * Most Recently Used and Priority based cache. A separate cache for each connection in the driver.
+ */
+
+/**
+   * Constructor.
+   *
+   * @param capacity Maximum capacity of the cache.
+   */
+
+function QueryContextCache (capacity) {
+  this.capacity = capacity;
+  this.idMap = new Map(); // Map for id and QCC
+  this.treeSet = new Set(); // Order data as per priority
+  this.priorityMap = new Map(); // Map for priority and QCC
+  this.newPriorityMap = new Map(); // Intermediate map for priority and QCC for current round of merging
+}
+
+/**
+   * Add an element in the cache.
+   *
+   * @param qce element to add
+   */
+QueryContextCache.prototype.sortTreeSet = function () {
+  this.treeSet = new Set(Array.from(this.treeSet).sort((a,b)=>a.getPriority()-b.getPriority()));
+};
+
+QueryContextCache.prototype.addQCE= function (qce) {
+  this.idMap.set(qce.id,qce);
+  this.newPriorityMap.set(qce.priority,qce);
+  this.treeSet.add(qce);
+  this.sortTreeSet();
+};
+
+/**
+   * Remove an element from the cache.
+   *
+   * @param qce element to remove.
+   */
+QueryContextCache.prototype.removeQCE = function (qce) {
+  this.idMap.delete(qce.id);
+  this.priorityMap.delete(qce.priority);
+  this.treeSet.delete(qce);
+};
+
+/**
+   * Replace the cache element with a new response element. Remove old element exist in the cache
+   * and add a new element received.
+   *
+   * @param oldQCE an element exist in the cache
+   * @param newQCE a new element just received.
+   */
+QueryContextCache.prototype.replaceQCE = function (oldQCE, newQCE) {
+  // Remove old element from the cache
+  this.removeQCE(oldQCE);
+  // Add new element in the cache
+  this.addQCE(newQCE);
+};
+
+/**
+   * Merge a new element comes from the server with the existing cache. Merge is based on read time
+   * stamp for the same id and based on priority for two different ids.
+   *
+   * @param id Database id.
+   * @param timestamp Last time read metadata from FDB.
+   * @param priority 0 to N number, where 0 is the highest priority. Eviction policy is based on
+   *     priority.
+   * @param context Opaque query context.
+   * 
+   * 
+   */
+QueryContextCache.prototype.merge = function (id, timestamp, priority, context) {
```

Comment:
why don't we receive here just new QCE instead of all its fields?

Reply:
we can do this even when we receive whole object (instead of separate fields) or when we receive the list - logic under the hood should not be different but it would be much more extensible when e.g. we add sth more to the query context element

Reply:
Fixed. From now on, we receive the qce element instead of the whole field in the merge function